File: tests/abi/preset_file_layout.cpp
#include "public.sdk/source/vst/vstpresetfile.h"

#include <cstring>
#include <cstdio>
#include <cstddef>

static void printChunkID (const char* label, const Steinberg::Vst::ChunkID& id)
{
	std::printf ("%s %.4s\n", label, id);
}

int main ()
{
	Steinberg::Vst::ChunkID header = {'V', 'S', 'T', '3'};
	Steinberg::Vst::ChunkID componentState = {'C', 'o', 'm', 'p'};
	Steinberg::Vst::ChunkID controllerState = {'C', 'o', 'n', 't'};
	Steinberg::Vst::ChunkID programData = {'P', 'r', 'o', 'g'};
	Steinberg::Vst::ChunkID metaInfo = {'I', 'n', 'f', 'o'};
	Steinberg::Vst::ChunkID chunkList = {'L', 'i', 's', 't'};
	std::printf ("sizeof.ChunkID %zu\n", sizeof (Steinberg::Vst::ChunkID));
	std::printf ("sizeof.Entry %zu\n", sizeof (Steinberg::Vst::PresetFile::Entry));
	std::printf ("alignof.Entry %zu\n", alignof (Steinberg::Vst::PresetFile::Entry));
	std::printf ("offsetof.Entry.id %zu\n", offsetof (Steinberg::Vst::PresetFile::Entry, id));
	std::printf ("offsetof.Entry.offset %zu\n", offsetof (Steinberg::Vst::PresetFile::Entry, offset));
	std::printf ("offsetof.Entry.size %zu\n", offsetof (Steinberg::Vst::PresetFile::Entry, size));
	std::printf ("PresetFile.kMaxEntries %d\n", 128);
	std::printf ("PresetFile.kFormatVersion %d\n", 1);
	std::printf ("PresetFile.kClassIDSize %d\n", 32);
	std::printf ("PresetFile.kHeaderSize %zu\n", sizeof (Steinberg::Vst::ChunkID) + sizeof (Steinberg::int32) + 32 + sizeof (Steinberg::TSize));
	std::printf ("PresetFile.kListOffsetPos %zu\n", sizeof (Steinberg::Vst::ChunkID) + sizeof (Steinberg::int32) + 32);
	std::printf ("ChunkType.kHeader %d\n", Steinberg::Vst::kHeader);
	std::printf ("ChunkType.kComponentState %d\n", Steinberg::Vst::kComponentState);
	std::printf ("ChunkType.kControllerState %d\n", Steinberg::Vst::kControllerState);
	std::printf ("ChunkType.kProgramData %d\n", Steinberg::Vst::kProgramData);
	std::printf ("ChunkType.kMetaInfo %d\n", Steinberg::Vst::kMetaInfo);
	std::printf ("ChunkType.kChunkList %d\n", Steinberg::Vst::kChunkList);
	std::printf ("ChunkType.kNumPresetChunks %d\n", Steinberg::Vst::kNumPresetChunks);
	printChunkID ("ChunkID.kHeader", header);
	printChunkID ("ChunkID.kComponentState", componentState);
	printChunkID ("ChunkID.kControllerState", controllerState);
	printChunkID ("ChunkID.kProgramData", programData);
	printChunkID ("ChunkID.kMetaInfo", metaInfo);
	printChunkID ("ChunkID.kChunkList", chunkList);
	std::printf ("isEqualID.header.header %d\n", std::memcmp (header, header, sizeof (Steinberg::Vst::ChunkID)) == 0);
	std::printf ("isEqualID.header.list %d\n", std::memcmp (header, chunkList, sizeof (Steinberg::Vst::ChunkID)) == 0);
	return 0;
}

File: tools/preset_file_layout.zig
const std = @import("std");
const preset_file = @import("zig-vst3").pluginterfaces.vst.vstpresetfile;

fn printChunkID(writer: anytype, label: []const u8, id: preset_file.ChunkID) !void {
    try writer.print("{s} {s}\n", .{ label, id });
}

pub fn main() !void {
    const stdout = std.fs.File.stdout().deprecatedWriter();
    try stdout.print("sizeof.ChunkID {}\n", .{@sizeOf(preset_file.ChunkID)});
    try stdout.print("sizeof.Entry {}\n", .{@sizeOf(preset_file.Entry)});
    try stdout.print("alignof.Entry {}\n", .{@alignOf(preset_file.Entry)});
    try stdout.print("offsetof.Entry.id {}\n", .{@offsetOf(preset_file.Entry, "id")});
    try stdout.print("offsetof.Entry.offset {}\n", .{@offsetOf(preset_file.Entry, "offset")});
    try stdout.print("offsetof.Entry.size {}\n", .{@offsetOf(preset_file.Entry, "size")});
    try stdout.print("PresetFile.kMaxEntries {}\n", .{preset_file.kMaxEntries});
    try stdout.print("PresetFile.kFormatVersion {}\n", .{preset_file.kFormatVersion});
    try stdout.print("PresetFile.kClassIDSize {}\n", .{preset_file.kClassIDSize});
    try stdout.print("PresetFile.kHeaderSize {}\n", .{preset_file.kHeaderSize});
    try stdout.print("PresetFile.kListOffsetPos {}\n", .{preset_file.kListOffsetPos});
    try stdout.print("ChunkType.kHeader {}\n", .{@intFromEnum(preset_file.ChunkType.kHeader)});
    try stdout.print("ChunkType.kComponentState {}\n", .{@intFromEnum(preset_file.ChunkType.kComponentState)});
    try stdout.print("ChunkType.kControllerState {}\n", .{@intFromEnum(preset_file.ChunkType.kControllerState)});
    try stdout.print("ChunkType.kProgramData {}\n", .{@intFromEnum(preset_file.ChunkType.kProgramData)});
    try stdout.print("ChunkType.kMetaInfo {}\n", .{@intFromEnum(preset_file.ChunkType.kMetaInfo)});
    try stdout.print("ChunkType.kChunkList {}\n", .{@intFromEnum(preset_file.ChunkType.kChunkList)});
    try stdout.print("ChunkType.kNumPresetChunks {}\n", .{@intFromEnum(preset_file.ChunkType.kNumPresetChunks)});
    try printChunkID(stdout, "ChunkID.kHeader", preset_file.getChunkID(.kHeader));
    try printChunkID(stdout, "ChunkID.kComponentState", preset_file.getChunkID(.kComponentState));
    try printChunkID(stdout, "ChunkID.kControllerState", preset_file.getChunkID(.kControllerState));
    try printChunkID(stdout, "ChunkID.kProgramData", preset_file.getChunkID(.kProgramData));
    try printChunkID(stdout, "ChunkID.kMetaInfo", preset_file.getChunkID(.kMetaInfo));
    try printChunkID(stdout, "ChunkID.kChunkList", preset_file.getChunkID(.kChunkList));
    try stdout.print("isEqualID.header.header {}\n", .{@intFromBool(preset_file.isEqualID(
        preset_file.getChunkID(.kHeader),
        preset_file.getChunkID(.kHeader),
    ))});
    try stdout.print("isEqualID.header.list {}\n", .{@intFromBool(preset_file.isEqualID(
        preset_file.getChunkID(.kHeader),
        preset_file.getChunkID(.kChunkList),
    ))});
}
